shopping-cart/shopping-cart-scala/shopping-cart/src/main/scala/com/example/shoppingcart/impl/ShoppingCartEntity.scala

```diff
@@ -214,114 +153,126 @@ object CheckedOut {
 
-object ShoppingCartException {
+object Rejected {
+  implicit val format: Format[Rejected] = Json.format
+}
 
-  /**
-    * Format for the ShoppingCartException.
-    *
-    * When a command fails, the error needs to be serialized and sent back to
-    * the node that requested it, this is used to do that.
-    */
-  implicit val format: Format[ShoppingCartException] = Json.format[ShoppingCartException]
+final case class CurrentState(state: ShoppingCartState) extends ShoppingCartReply
+
+object CurrentState {
+  implicit val format: Format[CurrentState] = Json.format
 }
 
 /**
-  * Akka serialization, used by both persistence and remoting, needs to have
-  * serializers registered for every type serialized or deserialized. While it's
-  * possible to use any serializer you want for Akka messages, out of the box
-  * Lagom provides support for JSON, via this registry abstraction.
-  *
-  * The serializers are registered here, and then provided to Lagom in the
-  * application loader.
-  */
+ * This is a marker trait for shopping cart commands.
+ * We will serialize them using Akka's Jackson support that is able to deal with the replyTo field.
+ * (see application.conf)
+ */
+trait ShoppingCartCommandSerializable
+
+/**
+ * This interface defines all the commands that the ShoppingCartEntity supports.
+ */
+sealed trait ShoppingCartCommand[R <: ShoppingCartReply] extends ExpectingReply[R] with ShoppingCartCommandSerializable
+
+/**
+ * A command to update an item.
+ *
+ * It has a reply type of [[Done]], which is sent back to the caller
+ * when all the events emitted by this command are successfully persisted.
+ */
+case class UpdateItem(productId: String, quantity: Int, replyTo: ActorRef[Confirmation])
+    extends ShoppingCartCommand[Confirmation]
+
+/**
+ * A command to get the current state of the shopping cart.
+ *
+ * The reply type is the ShoppingCart, and will contain the message to say to that
+ * person.
+ */
+case class Get(replyTo: ActorRef[CurrentState]) extends ShoppingCartCommand[CurrentState]
+
+/**
+ * A command to checkout the shopping cart.
+ *
+ * The reply type is the Done, which will be returned when the events have been
+ * emitted.
+ */
+case class Checkout(replyTo: ActorRef[Confirmation]) extends ShoppingCartCommand[Confirmation]
+
+/**
+ * Akka serialization, used by both persistence and remoting, needs to have
+ * serializers registered for every type serialized or deserialized. While it's
+ * possible to use any serializer you want for Akka messages, out of the box
+ * Lagom provides support for JSON, via this registry abstraction.
+ *
+ * The serializers are registered here, and then provided to Lagom in the
+ * application loader.
+ */
 object ShoppingCartSerializerRegistry extends JsonSerializerRegistry {
   override def serializers: Seq[JsonSerializer[_]] = Seq(
+    // state and events can use play-json, but commands should use jackson because of ActorRef[T] (see application.conf)
```

Comment:
Only events and states use play-json and Lagom's mechanism to declare programmatically serializers. Commands should use Akka's Jackson because of `ActorRef[Reply]`.